### PR TITLE
[API][Shop] Viewing only the enabled variants of a product fixes

### PR DESCRIPTION
**Pull request by TheMilek**
| Q               | A
| --------------- | -----
| Branch?         | master 
| Bug fix?        | no
| New feature?    | no
| BC breaks?      | no
| Related tickets | fixes to [PR](https://github.com/Sylius/Sylius/pull/13863)
| License         | MIT

This PR contains missing QueryItemExtension for ProductVariants that shows only enabled variants and corrects comments from [this](https://github.com/Sylius/Sylius/pull/13863) PR

### CODE REVIEW

File: src/Sylius/Bundle/ApiBundle/Doctrine/QueryItemExtension/EnabledProductVariantItemExtension.php [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+namespace Sylius\Bundle\ApiBundle\Doctrine\QueryItemExtension;
+
+use ApiPlatform\Core\Bridge\Doctrine\Orm\Extension\QueryItemExtensionInterface;
+use ApiPlatform\Core\Bridge\Doctrine\Orm\Util\QueryNameGeneratorInterface;
+use Doctrine\ORM\QueryBuilder;
+use Sylius\Bundle\ApiBundle\Context\UserContextInterface;
+use Sylius\Component\Core\Model\ProductVariantInterface;
+
+/** @experimental */
+final class EnabledProductVariantItemExtension implements QueryItemExtensionInterface
```
Comment (Rafikooo) [inline src/Sylius/Bundle/ApiBundle/Doctrine/QueryItemExtension/EnabledProductVariantItemExtension.php:23]: What about spec Class?

Comment (TheMilek) [reply src/Sylius/Bundle/ApiBundle/Doctrine/QueryItemExtension/EnabledProductVariantItemExtension.php:23]: You mean it's missing?
It's defined here:
https://github.com/Sylius/Sylius/pull/13877/files#diff-ea220efe1a03086966dba9a61b6d0e90b679269c6327afb1b056d47e33ec6636

File: src/Sylius/Bundle/ApiBundle/Doctrine/QueryItemExtension/EnabledProductVariantItemExtension.php [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+            return;
+        }
+
+        $user = $this->userContext->getUser();
+        if ($user !== null && in_array('ROLE_API_ACCESS', $user->getRoles(), true)) {
+            return;
+        }
+
+        $rootAlias = $queryBuilder->getRootAliases()[0];
+        $enabledParameterName = $queryNameGenerator->generateParameterName('enabled');
```
Comment (Rafikooo) [inline src/Sylius/Bundle/ApiBundle/Doctrine/QueryItemExtension/EnabledProductVariantItemExtension.php:47]: What do you think about `$enabled`?

**Review by Rafikooo — APPROVED** [review]

**Review by AdamKasp — CHANGES_REQUESTED** [review]
And missing tests, you should add tests for these changes.
- contract test with response `404` not found
- or/and simple behat with this behaviour (with tags `api` and `no-ui`

File: UPGRADE-API-1.12.md [hunk 3]
```diff
@@ -99,4 +99,4 @@ Here is how the response looks like:
 
 1. The 2nd parameter `localeCode` has been removed from `src/Sylius/Bundle/ApiBundle/Command/Cart/PickupCart.php` and now is set automatically by `src/Sylius/Bundle/ApiBundle/DataTransformer/LocaleCodeAwareInputCommandDataTransformer.php`.
 
-1. The `api/v2/admin/products` and `api/v2/admin/products/{code}` have been changed and the `defaultVariant` field has been removed.
+1. The responses of endpoints `/api/v2/admin/products` and `/api/v2/admin/products/{code}` have been changed in such a way that the field `defaultVariant` has been removed.
```
Comment (coldic3) [inline UPGRADE-API-1.12.md:102]: This line https://github.com/Sylius/Sylius/blob/efc9d781de28c18f9cd1ef778178e863257a1d58/src/Sylius/Bundle/ApiBundle/Swagger/ProductDocumentationNormalizer.php#L41 should also be deleted.

File: src/Sylius/Behat/Context/Api/Shop/ProductContext.php [hunk 4]
```diff
@@ -598,14 +598,13 @@ private function hasProductOptionWithNameAndValue(string $optionValue, string $o
...
-
-                foreach ($variants as $variantIri) {
-                    if ($this->variantHasProductOptionValue($variantIri, $optionValue))
-                    {
-                        return true;
-                    }
+            if (!$this->hasProductOptionWithName($optionIri, $optionName)) {
+                continue;
+            }
+            $variants = $this->responseChecker->getValue($response, 'variants');
```
Comment (coldic3) [inline src/Sylius/Behat/Context/Api/Shop/ProductContext.php:604]:
```suggestion
            }

            $variants = $this->responseChecker->getValue($response, 'variants');
```

File: src/Sylius/Bundle/ApiBundle/Doctrine/QueryItemExtension/EnabledProductVariantItemExtension.php [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+use Sylius\Component\Core\Model\ProductVariantInterface;
+
+/** @experimental */
+final class EnabledProductVariantItemExtension implements QueryItemExtensionInterface
+{
+    public function __construct(private UserContextInterface $userContext)
+    {
+    }
+
+    public function applyToItem(
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Doctrine/QueryItemExtension/EnabledProductVariantItemExtension.php:29]: You should add a contract test to check this behaviour

**Review by Ferror — APPROVED** [review]

File: features/product/viewing_product_variants/viewing_product_variant.feature [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+        Given the store operates on a channel named "Web-US" in "USD" currency
+        And the store has a "Super Cool T-Shirt" configurable product
+        And this product has "Small", "Medium" and "Large" variants
+        And the product "Super Cool T-Shirt" has also an "Extra Large" variant
+        And the "Extra Large" product variant is disabled
+
+    @api @no-ui
+    Scenario: Seeing the disabled variant is not possible
+        When I view variants
+        Then "Extra Large" variant should not be visible
```
Comment (GSadee) [inline features/product/viewing_product_variants/viewing_product_variant.feature:17]:
```suggestion
        Then the "Extra Large" variant should not be visible
```

File: features/product/viewing_product_variants/viewing_product_variant.feature [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+
+    @api @no-ui
+    Scenario: Seeing the enabled variant is possible
+        When I view variants
+        Then "Small" variant and "Medium" variant and "Large" variant should be visible
```
Comment (GSadee) [inline features/product/viewing_product_variants/viewing_product_variant.feature:22]:
```suggestion
        Then the "Small", "Medium" and "Large" variants should be visible
```

**Review by GSadee — APPROVED** [review]

**Review by AdamKasp — COMMENTED** [review]
do you check this comment? https://github.com/Sylius/Sylius/pull/13863#discussion_r849424516
because it is a resolved but without any implementation (or i missing sth)

**Review by AdamKasp — APPROVED** [review]

File: src/Sylius/Behat/Context/Api/Shop/ProductVariantContext.php [hunk 8]
```diff
@@ -252,6 +252,37 @@ public function variantAndVariantShouldNotBeDiscounted(ProductVariantInterface .
         }
     }
 
+    /**
+     * @Then I should not see :variant variant
+     */
+    public function iShouldNotSeeVariant($variant): void
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/ProductVariantContext.php:258]:
```suggestion
    public function iShouldNotSeeVariant(ProductVariantInterface $variant): void
```

File: src/Sylius/Behat/Context/Transform/ProductVariantContext.php [hunk 9]
```diff
@@ -95,6 +95,16 @@ public function getProductVariantByName($name)
         return $productVariants[0];
     }
 
+    /**
+     * @Transform /^"([^"]+)", "([^"]+)" and "([^"]+)" variants$/
+     */
+    public function getVariantsByNames(...$variantNames)
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Transform/ProductVariantContext.php:101]:
```suggestion
    public function getVariantsByNames(string ...$variantNames): array
```

**Review by GSadee — APPROVED** [review]

**Comment by GSadee** [comment]
Thank you, Kamil! :tada:

**Pull request merged**